### PR TITLE
[Lease] Refactor lease renew request via raft

### PR DESCRIPTION
**Pull request by ahrtr**
Previously the lease renew request can only be processed by the leader. When a follower receives the renew request, it just forwards the request to the leader via an internal http channel. This isn't accurate because the leader may change during the process.

When a leader receives the renew request, the previous implementation follows a three stages workflow: pre-raft, raft and post-raft. It's  too complicated and error prone, and the raft is more like just a network transport channel instead of a consensus mechanism in this case. 

So in this PR, we process the renew request via the raft directly, it can greatly simplify the code.

### CODE REVIEW

**Comment by ahrtr** [comment]
cc @serathius @ptabor @spzala @jpbetz @gyuho @xiang90 @jingyih @hexfusion @wenjiaswe @mitake 
     @wojtek-t @endocrimes @mgates @cosgroveb for feedback. Thanks!

**Comment by ahrtr** [comment]
Please note that I intentionally broke down the PR into three commits, and I recommend to review this PR commit by commit. 

[lease: add a lease_new request into raft_internal.proto](https://github.com/etcd-io/etcd/pull/14094/commits/23d98746a1ce25da9804061d6d7acf0ab8190663)
[lease: remove http handler for renew request](https://github.com/etcd-io/etcd/pull/14094/commits/31b19b01d967e39dcd56c65546bea741ef5e9d5f)
[lease: process renew request via raft](https://github.com/etcd-io/etcd/pull/14094/commits/49dcdc998beb0f1ddf32ca66bdd013cfa158f45b)

**Comment by serathius** [comment]
Discussion on lease issues and proposed fixes are so spread I find myself unable to comprehend it all. 

@ahrtr can you help me understand how this change refers to https://github.com/etcd-io/etcd/issues/13915 and what is our overall plan. It would be great to have one umbrella issue that shows high level plan so we can make sure we are going in right direction.

Not sure I remember ever discussing moving Grant to raft as there were some performance concerns. With a lot of leases, grant requests can be done very frequently and they are latency sensitive making them much more vulnerable to network hiccup. Before we proceed with this PR would be good to dig up the original discussion and do some loadtesting. I haven't look to deep into this PR so I might be wrong about performance concerns.

**Comment by ahrtr** [comment]
This is a standalone refactor on `Renew` instead of `Grant`. 

It took me sometime to go through some historical PRs ( [9924](https://github.com/etcd-io/etcd/pull/9924), [9526](https://github.com/etcd-io/etcd/pull/9526), [9699](https://github.com/etcd-io/etcd/pull/9699), [13508](https://github.com/etcd-io/etcd/pull/13508)  ) before delivering this PR. 

I have thought about three solutions (see below) to refactor lease, but eventually I realized that none of them are feasible.
1. persist expiryTime instead of remainingTTL into db, and it's similar to what [9526](https://github.com/etcd-io/etcd/pull/9526) did. And we don't need the Checkpoint functionality at all, each member, including leader and followers, calculate the expiryTime themselves.  Since different nodes may have inconsistent wall time, so only the leader can check the expiry of each lease. The reason why each follower also calcuates the expiryTime is to prepare for the leader change case. Once a follower becomes a leader, it just uses its local expiryTime to continue to serve the leases.  

This solution can greatly simplify the overall design and implementation, because we don't need the checkpoint functionality anymore.  If a member's time jumps forwards or backwards drastically, then we will run into issue no matter which solution we follow. But this solution will be even worse in this case, because existing implementation will only be impacted when the leader's time jumps; but this solution might run into issue if any member's time jumps. Of course, each time revoking & renewing can fix & reset the time jumps. 

Another downside of this solution is when the cluster is down for some time (such as maintenance window), then all leases might be expired when the cluster starts again. The existing implementation will not have this issue, because it persists the remainingTTL instead of expiryTime. The customer/client shouldn't pay for the server side issue. So persisting remainingTTL makes more sense then persisting expiryTime from this perspective. 

2. Introduce a concept `clusterClock`, so that all members, including  leader and followers, depend on the `clusterClock` instead of its local wal time. In this case, we don't need the checkpoint either. It can also greatly simplify the overall design & implementation.  But it isn't that easy to implement a clusterClock. :(. Please refer to [logcabin](https://ongardie.net/blog/logcabin-2015-02-27/). Please also let me know if you have any proposals. :) 

3. We can introduce NTP (network time protocol), but it will complicate the overall design. And it still can't guarantee the consistent of wall time between members .

Speaking to the overall plan, my thoughts are:
1. Keep the current implementation of `Grant`, `Revoke`, `Checkpoint` as they are for now. All of them are going through raft, it should be OK. Especially the checkpoint syncs remainingTTL to followers via raft, which is good to me. Please see comment (`remainingTTL` vs `expiryTime`) above on the first infeasible solution.
2. Refactor lease renew request via raft. This is exactly this PR fixes. Please read the description of this PR. 
5. Modify LeaseLease/ListLease to make sure the member has applied the latest commitID. We just need to use [linearizableReadNotify](https://github.com/etcd-io/etcd/blob/29b964df25694cdb1e0039531ef213743e04e619/server/etcdserver/v3_server.go#L883). I think the PR [13882](https://github.com/etcd-io/etcd/pull/13882) should be the right direction. 
6. Regarding to the [LeaseTimeToLive](https://github.com/etcd-io/etcd/blob/29b964df25694cdb1e0039531ef213743e04e619/server/etcdserver/v3_server.go#L320), let's keep it as it's for now, b ecause we can only depend on the leader's local wal time. The followers' local time may be inconsistency.  

So in summary, we only need to refactor renew (this PR) and modify `Listlease` ( [13882](https://github.com/etcd-io/etcd/pull/13882) ). They are two separate & independent changes.

File: server/etcdserver/api/etcdhttp/peer.go [hunk 1]
```diff
@@ -63,7 +63,6 @@ func newPeerHandler(
 	mux.Handle(peerMembersPath, peerMembersHandler)
 	mux.Handle(peerMemberPromotePrefix, peerMemberPromoteHandler)
 	if leaseHandler != nil {
-		mux.Handle(leasehttp.LeasePrefix, leaseHandler)
```
[the remark this answers is not in the document]
Comment (ahrtr) [reply server/etcdserver/api/etcdhttp/peer.go:66]: There isn't any change on the client facing API. This is just for peer communication. But there may be some errors during upgrade when only the leader has already been upgraded.  The renew request to the leader may get a 404 response. 

Once all of them reach to the same version, everything will be working well.

[the remark this answers is not in the document]
Comment (ahrtr) [reply server/etcdserver/api/etcdhttp/peer.go:66]: Yes, you are right. It's even complicated than what we have thought about. I will provide a more detailed plan on how to smoothly support this, which should be backward & forward compatible. 

Actually the PR [9924](https://github.com/etcd-io/etcd/pull/9924) is a bad example, because it breaks the compatibility when part of the node in 3.5 during upgrading etcd from 3.4 to 3.5. But since the the [checkpoint](https://github.com/etcd-io/etcd/blob/5a7df035929a52edabb8130c63926e6531a371c3/server/etcdmain/config.go#L261) is an experimental feature, and it's disabled by default, so nobody runs into the upgrade compatibility issue so far.

[the remark this answers is not in the document]
Comment (serathius) [reply server/etcdserver/api/etcdhttp/peer.go:66]: https://github.com/etcd-io/etcd/pull/9924 is an interesting case as it's etcd command line configuration that affects etcd state. As you cannot imminently set this flags on all cluster members, it means that different members will have different state. For leases this is not a big deal as this is transient, however we should be careful about such changes. 

Depending on number of such situations we might want to consider having such configuration negotiated via raft. I wanted to propose it some time ago, however this is currently not high on list of my priories, as we have a lot of other problems.

[the remark this answers is not in the document]
Comment (ahrtr) [reply server/etcdserver/api/etcdhttp/peer.go:66]: > Depending on number of such situations we might want to consider having such configuration negotiated via raft. 

It looks like a good & generic point. We can discuss it separately if needed.

[the remark this answers is not in the document]
Comment (ahrtr) [reply server/etcdserver/api/etcdhttp/peer.go]: I added a new flag `--experimental-enable-lease-v2-renew`, and the default value is `true` in 3.6. It means that etcd will continue to use the legacy renew implementation by default. We can consider to deprecate the legacy renew implementation in 3.7 and set the flag as `false` by default. Eventually we can remove the legacy renew implementation in 3.8.

File: server/etcdserver/apply/apply.go [hunk 2]
```diff
@@ -207,6 +208,11 @@ func (a *applierV3backend) LeaseRevoke(lc *pb.LeaseRevokeRequest) (*pb.LeaseRevo
 	return &pb.LeaseRevokeResponse{Header: a.newHeader()}, err
 }
 
+func (a *applierV3backend) LeaseRenew(lc *pb.LeaseKeepAliveRequest) (*pb.LeaseKeepAliveResponse, error) {
```
Comment (mitake) [inline server/etcdserver/apply/apply.go:211]: I think it's still possible to apply `LeaseRenew` entries issued by a stale leader unless it doesn't have a mechanism like comparing term https://github.com/etcd-io/etcd/issues/15247#issuecomment-1510362310 ?
I think all Raft messages issued by etcd itself might have similar problems potentially.
